src/NUnitFramework/framework/Constraints/Comparers/ComparisonState.cs

```diff
@@ -0,0 +1,62 @@
+using System.Collections.Generic;
+
+namespace NUnit.Framework.Constraints.Comparers
+{
+    internal struct ComparisonState
+    {
+        /// <summary>
+        /// Flag indicating whether or not this is the top level comparison.
+        /// </summary>
+        public readonly bool TopLevelComparison;
+
+        /// <summary>
+        /// A list of tracked comparisons
+        /// </summary>
+        private readonly Comparison[] _comparisons;
+
+        public ComparisonState(bool topLevelComparison)
+        {
+            TopLevelComparison = topLevelComparison;
+            _comparisons = new Comparison[0];
+        }
+
+        private ComparisonState(bool topLevelComparison, Comparison[] comparisons)
+        {
+            TopLevelComparison = topLevelComparison;
+            _comparisons = comparisons;
+        }
+
+        public ComparisonState PushComparison(bool topLevelComparison, object x, object y)
+        {
+            var comparisons = new Comparison[_comparisons.Length + 1];
+            _comparisons.CopyTo(comparisons, 0);
+            comparisons[_comparisons.Length] = new Comparison(x, y);
+
+            return new ComparisonState(
+                topLevelComparison,
+                comparisons
+            );
+        }
+
+        public bool DidCompare(object x, object y)
+        {
+            for(var i = 0; i < _comparisons.Length; i++)
+                if (_comparisons[i].X == x && _comparisons[i].Y == y)
```

Comment:
Maybe I'm missing something, but since we are only looking for recursion, wouldn't it be more efficient here to only test for reference equality, `Object.ReferenceEquals(_comparisons[0].X, x)`?

Reply:
The `==` operator checks reference equality if one of the operands is a reference type. It can't be overloaded in this scenario since both operands are `object`. `ReferenceEquals` might be clearer though?

Reply:
(The reference equality of `==` on `object` is also a pitfall. `obj == "foo"` is checking reference equality, which is different than `obj as string == "foo"` which uses the `==(string, string)` overload.)

Reply:
Thanks for the suggestion! I'll try and get this done as soon as possible.

Reply:
Converted to `Object.ReferenceEquals()`